### ISSUE TITLE
Recovering a nextcloud setup

### ISSUE THREAD

**Issue by mbewley** [nextcloud-snap/nextcloud-snap#1992]
Hi, I've been using nextcloud since it used to be called owncloud. I've been running a nextcloud server at home using a snap on Ubuntu for a few years now, as a nice easy way to manage it. Recently, it got stuck in a strange state - snaps weren't mounting properly, so there was nothing in the /snap/nextcloud directory. With some fiddling and restarting, the snaps started mounting again. However, the "nextcloud.occ" binary was missing, even though the snap no longer complained it was broken.
So... I did a "snap revert nextcloud", which worked successfully (from 22.2.3snap3 to 22.2.3snap2). Nextcloud.occ now exists!
Next problem... any call to nextcloud.occ sits spinning on "Waiting for PHP..." endlessly. Trying to upgrade again (snap refresh nextcloud) sits on this for 10 minutes and then dies:

- Run pre-refresh hook of "nextcloud" snap if present (run hook "pre-refresh": 
-----
Waiting for Apache...

Help! I have my separate data directory safe, but don't want to do any destructive reinstalling in case I lose things like file history and other database relevant stuff. Any ideas what could be wrong?

**Comment by mbewley** [comment]
Update - running "snap services nextcloud", most of the services were listed as "inactive", even after a "snap restart nextcloud". So I tried manually enabling each of them via e.g. "snap start nextcloud.apache". This worked, and I can now log into my nextcloud instance again. Next step some careful additional backing up, then attempting to refresh to the latest snap again...

**Comment by kyrofa** [comment]
Huh, wonder what happened to snapd in there. Nice detective work, though. Any luck updating to the latest snap? Do those services fire up as they should if you reboot? Or do you need to start them manually every time, now?

**Comment by mbewley** [comment]
I rebooted the machine, and the services stayed down:

Service                    Startup   Current   Notes
nextcloud.apache           disabled  inactive  -
nextcloud.logrotate        enabled   inactive  timer-activated
nextcloud.mdns-publisher   disabled  inactive  -
nextcloud.mysql            disabled  inactive  -
nextcloud.nextcloud-cron   disabled  inactive  -
nextcloud.nextcloud-fixer  disabled  inactive  -
nextcloud.php-fpm          disabled  inactive  -
nextcloud.redis-server     disabled  inactive  -
nextcloud.renew-certs      disabled  inactive  -


Manually started each of them again, and nextcloud started working again...
Then I did a "snap refresh nextcloud" to see if the upgrade would work.

After the update (which worked), the services were all stopped again as above. Manually starting as above worked.
On a final restart with the refreshed snap, the services still failed to come up.

Where should I look to see what's going on? Presumably it's attempting to start these sub services when the nextcloud snap starts, but failing?

**Comment by kyrofa** [comment]
> Where should I look to see what's going on?

I'm afraid I'm uncertain, that has never happened to me before. It sounds like snapd is having troubles, I suggest starting a thread on the [snapcraft forum](https://forum.snapcraft.io/) to see if anyone there has some useful insights.

Out of curiosity, can I see the output of `snap list nextcloud --all`, please?

> Presumably it's attempting to start these sub services when the nextcloud snap starts, but failing?

It's possible. You should be able to see such failures in the logs though. Try `journalctl -uf snap.nextcloud.apache` for example.

**Comment by mbewley** [comment]
`snap list nextcloud --all` gives:

Name       Version      Rev    Tracking       Publisher   Notes
nextcloud  22.2.3snap2  28713  latest/stable  nextcloud✓  disabled
nextcloud  22.2.3snap3  29223  latest/stable  nextcloud✓  -

**Comment by mbewley** [comment]
Ok, I've finally got my backups sorted properly. The services still aren't coming up on reboot. If I do a `sudo snap start nextcloud`, you get the following logs from `journalctl -u snap.nextcloud.apache`:

-- Reboot --
Feb 13 08:59:59 ant systemd[1]: Started Service for snap application nextcloud.apache.
Feb 13 08:59:59 ant nextcloud.apache[8293]: cannot change profile for the next exec call: No such file or directory
Feb 13 08:59:59 ant nextcloud.apache[8144]: snap-update-ns failed with code 1
Feb 13 08:59:59 ant systemd[1]: snap.nextcloud.apache.service: Main process exited, code=exited, status=1/FAILURE
Feb 13 08:59:59 ant systemd[1]: snap.nextcloud.apache.service: Failed with result 'exit-code'.
Feb 13 08:59:59 ant systemd[1]: snap.nextcloud.apache.service: Scheduled restart job, restart counter is at 1.
Feb 13 08:59:59 ant systemd[1]: Stopped Service for snap application nextcloud.apache.
Feb 13 08:59:59 ant systemd[1]: Started Service for snap application nextcloud.apache.
Feb 13 08:59:59 ant nextcloud.apache[8415]: cannot change profile for the next exec call: No such file or directory
Feb 13 08:59:59 ant nextcloud.apache[8342]: snap-update-ns failed with code 1
Feb 13 08:59:59 ant systemd[1]: snap.nextcloud.apache.service: Main process exited, code=exited, status=1/FAILURE
Feb 13 08:59:59 ant systemd[1]: snap.nextcloud.apache.service: Failed with result 'exit-code'.
Feb 13 08:59:59 ant systemd[1]: snap.nextcloud.apache.service: Scheduled restart job, restart counter is at 2.
Feb 13 08:59:59 ant systemd[1]: Stopped Service for snap application nextcloud.apache.
Feb 13 08:59:59 ant systemd[1]: Started Service for snap application nextcloud.apache.
Feb 13 08:59:59 ant nextcloud.apache[8588]: cannot change profile for the next exec call: No such file or directory
Feb 13 08:59:59 ant nextcloud.apache[8486]: snap-update-ns failed with code 1
Feb 13 08:59:59 ant systemd[1]: snap.nextcloud.apache.service: Main process exited, code=exited, status=1/FAILURE
Feb 13 08:59:59 ant systemd[1]: snap.nextcloud.apache.service: Failed with result 'exit-code'.
Feb 13 08:59:59 ant systemd[1]: snap.nextcloud.apache.service: Scheduled restart job, restart counter is at 3.
Feb 13 08:59:59 ant systemd[1]: Stopped Service for snap application nextcloud.apache.
Feb 13 08:59:59 ant systemd[1]: Started Service for snap application nextcloud.apache.
Feb 13 09:00:00 ant nextcloud.apache[8693]: cannot change profile for the next exec call: No such file or directory
Feb 13 09:00:00 ant nextcloud.apache[8632]: snap-update-ns failed with code 1
Feb 13 09:00:00 ant systemd[1]: snap.nextcloud.apache.service: Main process exited, code=exited, status=1/FAILURE
Feb 13 09:00:00 ant systemd[1]: snap.nextcloud.apache.service: Failed with result 'exit-code'.
Feb 13 09:00:00 ant systemd[1]: snap.nextcloud.apache.service: Scheduled restart job, restart counter is at 4.
Feb 13 09:00:00 ant systemd[1]: Stopped Service for snap application nextcloud.apache.
Feb 13 09:00:00 ant systemd[1]: Started Service for snap application nextcloud.apache.
Feb 13 09:00:00 ant nextcloud.apache[8819]: cannot change profile for the next exec call: No such file or directory
Feb 13 09:00:00 ant nextcloud.apache[8764]: snap-update-ns failed with code 1
Feb 13 09:00:00 ant systemd[1]: snap.nextcloud.apache.service: Main process exited, code=exited, status=1/FAILURE
Feb 13 09:00:00 ant systemd[1]: snap.nextcloud.apache.service: Failed with result 'exit-code'.
Feb 13 09:00:00 ant systemd[1]: snap.nextcloud.apache.service: Scheduled restart job, restart counter is at 5.
Feb 13 09:00:00 ant systemd[1]: Stopped Service for snap application nextcloud.apache.
Feb 13 09:00:00 ant systemd[1]: snap.nextcloud.apache.service: Start request repeated too quickly.
Feb 13 09:00:00 ant systemd[1]: snap.nextcloud.apache.service: Failed with result 'exit-code'.
Feb 13 09:00:00 ant systemd[1]: Failed to start Service for snap application nextcloud.apache.



Any ideas?

**Comment by mbewley** [comment]
Running just `journalctl -r`, (so I get all results and timestamps are in reverse order), I get this on starting nextcloud. Lots of the same error messages about APPARMOR.


Feb 13 09:06:11 ant nextcloud.nextcloud-fixer[10883]: cannot change profile for the next exec call: No such file or directory
Feb 13 09:06:11 ant audit[10883]: AVC apparmor="DENIED" operation="change_onexec" info="label not found" error=-2 profile="unconfined" name="snap-update-ns.nextcloud" pid=10883 comm="snap-confine"
Feb 13 09:06:11 ant systemd[1]: snap.nextcloud.php-fpm.service: Failed with result 'exit-code'.
Feb 13 09:06:11 ant systemd[6740]: tmp-snap.rootfs_GUdDsg.mount: Succeeded.
Feb 13 09:06:11 ant systemd[1]: tmp-snap.rootfs_GUdDsg.mount: Succeeded.
Feb 13 09:06:11 ant systemd[2519]: tmp-snap.rootfs_ZOUlzS.mount: Succeeded.
Feb 13 09:06:11 ant systemd[6740]: tmp-snap.rootfs_ZOUlzS.mount: Succeeded.
Feb 13 09:06:11 ant systemd[1]: snap.nextcloud.php-fpm.service: Main process exited, code=exited, status=1/FAILURE
Feb 13 09:06:11 ant nextcloud.php-fpm[10800]: snap-update-ns failed with code 1
Feb 13 09:06:11 ant kernel: audit: type=1400 audit(1644703571.356:52): apparmor="DENIED" operation="change_onexec" info="label not found" error=-2 profile="unconfined" name="snap-update-ns.nextcloud" pid=10881 >
Feb 13 09:06:11 ant nextcloud.php-fpm[10881]: cannot change profile for the next exec call: No such file or directory
Feb 13 09:06:11 ant audit[10881]: AVC apparmor="DENIED" operation="change_onexec" info="label not found" error=-2 profile="unconfined" name="snap-update-ns.nextcloud" pid=10881 comm="snap-confine"
Feb 13 09:06:11 ant systemd[1]: tmp-snap.rootfs_ZOUlzS.mount: Succeeded.
Feb 13 09:06:11 ant sudo[10675]: pam_unix(sudo:session): session closed for user root
Feb 13 09:06:11 ant systemd[1]: snap.nextcloud.mdns-publisher.service: Failed with result 'exit-code'.
Feb 13 09:06:11 ant systemd[2519]: tmp-snap.rootfs_DlGsjB.mount: Succeeded.
Feb 13 09:06:11 ant systemd[1]: snap.nextcloud.mdns-publisher.service: Main process exited, code=exited, status=1/FAILURE
Feb 13 09:06:11 ant systemd[6740]: tmp-snap.rootfs_DlGsjB.mount: Succeeded.
Feb 13 09:06:11 ant kernel: audit: type=1400 audit(1644703571.332:51): apparmor="DENIED" operation="change_onexec" info="label not found" error=-2 profile="unconfined" name="snap-update-ns.nextcloud" pid=10877 >
Feb 13 09:06:11 ant nextcloud.mdns-publisher[10744]: snap-update-ns failed with code 1
Feb 13 09:06:11 ant nextcloud.mdns-publisher[10877]: cannot change profile for the next exec call: No such file or directory
Feb 13 09:06:11 ant audit[10877]: AVC apparmor="DENIED" operation="change_onexec" info="label not found" error=-2 profile="unconfined" name="snap-update-ns.nextcloud" pid=10877 comm="snap-confine"
Feb 13 09:06:11 ant systemd[1]: tmp-snap.rootfs_DlGsjB.mount: Succeeded.
Feb 13 09:06:11 ant systemd[1]: snap.nextcloud.apache.service: Failed with result 'exit-code'.
Feb 13 09:06:11 ant systemd[1]: snap.nextcloud.apache.service: Main process exited, code=exited, status=1/FAILURE
Feb 13 09:06:11 ant nextcloud.apache[10724]: snap-update-ns failed with code 1
Feb 13 09:06:11 ant kernel: audit: type=1400 audit(1644703571.284:50): apparmor="DENIED" operation="change_onexec" info="label not found" error=-2 profile="unconfined" name="snap-update-ns.nextcloud" pid=10858 >
Feb 13 09:06:11 ant kernel: kauditd_printk_skb: 30 callbacks suppressed
Feb 13 09:06:11 ant nextcloud.apache[10858]: cannot change profile for the next exec call: No such file or directory
Feb 13 09:06:11 ant audit[10858]: AVC apparmor="DENIED" operation="change_onexec" info="label not found" error=-2 profile="unconfined" name="snap-update-ns.nextcloud" pid=10858 comm="snap-confine"
Feb 13 09:06:11 ant systemd[6740]: tmp-snap.rootfs_azVWX7.mount: Succeeded.
Feb 13 09:06:11 ant systemd[1]: tmp-snap.rootfs_azVWX7.mount: Succeeded.
Feb 13 09:06:11 ant systemd[2519]: tmp-snap.rootfs_azVWX7.mount: Succeeded.
Feb 13 09:06:11 ant systemd[1]: Started Service for snap application nextcloud.renew-certs.
Feb 13 09:06:11 ant systemd[1]: Started Service for snap application nextcloud.redis-server.
Feb 13 09:06:11 ant systemd[1]: Started Service for snap application nextcloud.php-fpm.
Feb 13 09:06:11 ant systemd[1]: Started Service for snap application nextcloud.nextcloud-fixer.
Feb 13 09:06:11 ant systemd[1]: Started Service for snap application nextcloud.nextcloud-cron.
Feb 13 09:06:11 ant systemd[1]: Started Service for snap application nextcloud.mysql.
Feb 13 09:06:11 ant systemd[1]: Started Service for snap application nextcloud.mdns-publisher.
Feb 13 09:06:11 ant systemd[1]: Started Service for snap application nextcloud.apache.
Feb 13 09:06:11 ant systemd[1]: Started Timer logrotate for snap application nextcloud.logrotate.
Feb 13 09:06:10 ant systemd[1]: Reloading.
Feb 13 09:06:10 ant sudo[10675]: pam_unix(sudo:session): session opened for user root by (uid=0)
Feb 13 09:06:10 ant sudo[10675]:  mbewley : TTY=pts/0 ; PWD=/home/mbewley ; USER=root ; COMMAND=/usr/bin/snap start nextcloud

**Comment by mbewley** [comment]
I think I've managed to fix it. Somehow under `sudo snap services`, a bunch of the nextcloud services had their "startup" mode set to disabled (!). So instead of just `sudo service start nextcloud.apache`, I did `sudo snap start --enable nextcloud.apache`, and it seems to work after a restart now, without manual intervention. I also tried disabling apparmor and then reenabling it, but I don't think that's what fixed it.
